### PR TITLE
Support Clojure v1.10, nREPL 0.5.3, enable nrepl connections from editors

### PR DESCRIPTION
**Pull request by klausharbo**
I have done some work to upgrade clojupyter: 

* support Clojure 1.9 and 1.10
* use nrepl 0.5.3 and cider-nrepl 0.20.0
* change the nrepl middleware to enable connecting to the nrepl server from code editors
* refactor the code quite a bit significantly reducing code redundancy
* updated `hmtl-demo.ipynb` to match the new version and changed it to use Vega Lite for the graphics example (the Highcharts example does not appear to work, as far as I can determine it is the case both for the existing `master`based on Clojure 1.8 and the updated version(?))

There's a [Changelog.md](https://github.com/klausharbo/clojupyter/blob/clojure-1.10-nrepl-0.5.3/Changelog.md) detaling the changes.

Let me know what you think.

-Klaus.

### CODE REVIEW

**Comment by roryk** [comment]
Thanks for all of this work, it looks like Travis is having a hard time finding the clojure executable, but I'm not super sure why.

**Comment by klausharbo** [comment]
Yes, I took a quick look at it, but couldn't see what is wrong.  The tests pass on my machine.  The Docker image `kxxoling/jupyter-clojure-docker` produces similar error, but since I have hardly any experience with Docker I don't really know what to look for...

**Comment by luojieshun** [comment]
Probably try update the the 

.travis.yml

**Comment by roryk** [comment]
I re-tried the master branch and it builds fine, so its something specific to this pull request that is causing the issue.

**Comment by klausharbo** [comment]
I found an error in test/core_test.clj - fixed on the branch.  I'll leave the branch alone except to find out why regression tests fail.

**Comment by klausharbo** [comment]
I got a Travis account and set it up to monitor the branch.  Still fails.

Although I am not sure what runtime Travis uses for its testing, I cloned the [jupyter-clojure-docker](kxxoling/jupyter-clojure-docker) image and made a [change](https://github.com/klausharbo/jupyter-clojure-docker/commit/42eb3e746438920f7bad66525958028dc22aa698) to manually install Clojure 1.10.  Resulting Dockerfile [here](https://github.com/klausharbo/jupyter-clojure-docker/blob/clojure-1.10-nrepl-0.5.3/Dockerfile).

Running `lein midje` in the resulting Docker image passes all tests.

I think the problem may have to do with Leiningen: The [Travis documentation says it uses Leiningen 2.4.x](https://docs.travis-ci.com/user/languages/clojure/#ci-environment-for-clojure-projects) and [NREPL says it needs 2.8.2+](https://nrepl.org/nrepl/0.5.3/installation.html).  Could that be it?

-Klaus.

**Comment by klausharbo** [comment]
The test now passes on [`klausharbo:clojure-1.10-nrepl-0.5.3`](https://github.com/klausharbo/clojupyter/tree/clojure-1.10-nrepl-0.5.3) in my Travis account.

It appears that the Travis setup does not currently support deps.edn/lein-tools-deps, moving dependencies back into `project.clj` and eliminating `deps.edn`did the trick.

Hopefuuly the pull request will show up as passing as well soon...

-K.

**Comment by roryk** [comment]
Thanks, this looks super good to me, thanks for all the work getting us up to date. Pinged @luojieshun to double-check; I haven't been writing any Clojure code lately.

**Comment by roryk** [comment]
Should get merged with a merge squash though to clean up the commit history.

File: project.clj [hunk 1]
```diff
@@ -1,31 +1,33 @@
-(defproject clojupyter "0.2.0-SNAPSHOT"
+(defproject clojupyter "0.1.1-SNAPSHOT"
```
Comment (roryk) [inline project.clj]: I think this should be 0.2.1, right?

File: test/clojupyter/core_test.clj [hunk 2]
```diff
@@ -60,165 +66,115 @@
...
+           {:status "ok"}
+           {:execution_state "idle"}]
+       (-> (get @zmq-messages 1) rcv (get-in [:content])
+           ((juxt :protocol_version :status :implementation
+                  (rcomp :language_info :name)
+                  (rcomp :language_info :mimetype)
+                  (rcomp :language_info :file_extension)
+                  (rcomp :language_info :version))))
+       => [messages/PROTOCOL-VERSION "ok" "clojupyter" "clojure" "text/x-clojure" ".clj"
+           (apply format "%d.%d.%d" ((juxt :major :minor :incremental) *clojure-version*))]))
```
Comment (roryk) [inline test/clojupyter/core_test.clj]: And I think here, we need to pull in the clojupyter version to get the kernel version, not the clojure version.

File: project.clj (hunk 1 above)
Comment (klausharbo) [reply project.clj]: Yes, you're right.

File: test/clojupyter/core_test.clj (hunk 2 above)
Comment (klausharbo) [reply test/clojupyter/core_test.clj]: No, I think it's correct.  The whole form is 

```
(-> (get @zmq-messages 1) rcv (get-in [:content])
    ((juxt :protocol_version :status :implementation
           (rcomp :language_info :name)
           (rcomp :language_info :mimetype)
           (rcomp :language_info :file_extension)
           (rcomp :language_info :version))))
=> [messages/PROTOCOL-VERSION "ok" "clojupyter" "clojure" "text/x-clojure" ".clj"
    (apply format "%d.%d.%d" ((juxt :major :minor :incremental) *clojure-version*))]
```
and `:protocol-version` lines up with `PROTOCOL-VERSION`, `:status` with `"ok"` and so on sequentially down the vector (`juxt` maintains order).

**Comment by klausharbo** [comment]
I am not sure about the protocol here: Should I do the corrections and commit squashing or will one of you guys do it on the merge?  I'm happy to do it, but if you'd normally do it that's fine with me also.

File: test/clojupyter/core_test.clj (hunk 2 above)
Comment (klausharbo) [reply test/clojupyter/core_test.clj]: Or, maybe I misunderstand: `:implementation` currently gives `"clojupyter"`, I'm pretty sure I didn't change that.  Are you saying it should be something like `"clojupyter-0.2.1-SNAPSHOT"`instead?

**Comment by roryk** [comment]
Heya, I can squash them when I do the merge-- you can do it as well if you'd like, either way is fine.

**Comment by klausharbo** [comment]
Please do go ahead.

File: project.clj (hunk 1 above)
Comment (klausharbo) [reply project.clj]: Made the change

**Comment by behrica** [comment]
Looking forward to this change.
I definitely will try it, once it is merged.

Writing the complex parts of the code in Emacs, while using the clojupyter notebook to "only contain minimal code to render render the results" offers great opportunities in my view.

**Comment by klausharbo** [comment]
This change will let you connect to a running kernel from a code editor such as CIDER/Emacs or Cursive using nREPL.  I think what you mean is to be able to connect the kernel to a running JVM using nREPL.  Personally I think I would prefer to work that way, as I always have a JVM and only sometimes want to use Jupyter to work with it.

Unfortunately implementing that is a bit more involved and not so is possible with this edition of the code.    It is something I'm interested in looking at, but it is not quite at the top of my list.

-K.

**Comment by roryk** [comment]
Thank you! I squashed and merged this in. I stole the last description you made to better describe what this pull request/commit is doing.

I added you as a collaborator-- the way we've been working is if a pull request is merged, then the person submitting it gets commit access. This has worked out awesome so far. Usually the way we've been working is fixing bugs we just commit without getting opinions from anyone else with a message about what bug its fixing and why, etc. Bigger features we've been drafting in a separate branch and then opening a pull request and requesting review to get some opinions from other folks using it.

**Comment by behrica** [comment]
> This change will let you connect to a running kernel from a code editor such as CIDER/Emacs or Cursive using nREPL. I think what you mean is to be able to connect the kernel to a running JVM using nREPL. Personally I think I would prefer to work that way, as I always have a JVM and only sometimes want to use Jupyter to work with it.
> 
> Unfortunately implementing that is a bit more involved and not so is possible with this edition of the code. It is something I'm interested in looking at, but it is not quite at the top of my list.
> 
> -K.

@klausharbo ,it works great for my use case.
I first start the notebook and the connect Cider to it.

Being in the same name spacewith both makes a very smooth workflow between Cider and the notebook.

I found it specialy usefull to write and test the code in Cider and only when working to copy paste it to the notebook.

**Pull request merged**